### ISSUE TITLE
Problem with $GLOBALS['NEWS_FILTER_EXCLUDE']

### ISSUE THREAD

**Issue by fritzmg** [codefog/contao-news_categories#57]
[ModuleNewsList.php#L83](https://github.com/codefog/contao-news_categories/blob/master/modules/ModuleNewsList.php#L83) does the following:
```php
$GLOBALS['NEWS_FILTER_EXCLUDE']    = array($news->id);
```
This excludes the current news entry (if available) in any following queries via
```php
\NewsModel::filterByCategories(…);
```
This is done to not include the current news entry on the reader page in a "related" news list, if present. However, this implementation is a bit problematic.

On a website, I am displaying several different news lists (filtered by certain categories) in the layout (mostly inserted via insert tag). These lists should always display all their news entries of course. However, on the news reader page I have also included a "related" news list.

Due to the aforementioned code, the current news entry gets removed from __all__ the news lists, not just from the "related" news list (at least if they are generated _after_ the "related" news list).

**Comment by qzminski** [comment]
This is a problem indeed, but I am afraid it cannot be fixed with the current implementation of the extension.

**Comment by fritzmg** [comment]
Hm, isn't it safe to assume, that after `generateRelated` has been called, the next call of `\NewsModel::filterByCategories(…);` will definitely be by the same module, thus you can simply do `unset( $GLOBALS['NEWS_FILTER_EXCLUDE'] );`? i.e.
```php
        // Exclude particular news items
        if (is_array($GLOBALS['NEWS_FILTER_EXCLUDE']) && !empty($GLOBALS['NEWS_FILTER_EXCLUDE'])) {
            $arrColumns[] = "$t.id NOT IN (" . implode(',', array_map('intval', $GLOBALS['NEWS_FILTER_EXCLUDE'])) . ")";
        unset( $GLOBALS['NEWS_FILTER_EXCLUDE'] );
        }
```

**Comment by qzminski** [comment]
I'm afraid you can't do that because the reset would need to be done in the news model class. Thus, you wouldn't know when actually to reset it - in count or find method?

**Comment by fritzmg** [comment]
Hm, true. I don't have another idea either :)

**Comment by qzminski** [comment]
I'll keep it open for consideration when I rewrite the extension for Contao 4.

**Comment by fritzmg** [comment]
For anyone else stumbling over this: currently I cicumvent this issue by inserting the related news list per insert tag, so that the module is processed after the others. However, this means that your other news list modules must not be inserted by insert tag either (or at least, they must be inserted somewhere _before_ the related news list).

**Comment by fritzmg** [comment]
Wait, isn't this rather easy to fix?
```php
…

$result = parent::generate();

unset( $GLOBALS['NEWS_FILTER_EXCLUDE'] );
unset( $GLOBALS['NEWS_FILTER_DEFAULT'] );
unset( $GLOBALS['NEWS_FILTER_CATEGORIES'] );

return $result;
```
In ModuleNewsList.php, generateRelated.

**Comment by qzminski** [comment]
Seems reasonable. Have you tried it?

**Comment by fritzmg** [comment]
Yep, I am using it in a project currently. Seems to work fine, can't think of any drawbacks.

**Comment by qzminski** [comment]
I will implement it then :+1:

**Comment by fritzmg** [comment]
Great :). Maybe do it in the regular generate method too, so make sure the globals don't "bleed".

**Comment by qzminski** [comment]
Fixed in fc106ae.
